File: CIMenu/StatusBarItemController.rb
class StatusBarItemController
  attr_accessor :image, :alternateImage, :menuIsActive

  def initialize
    @statusBar = NSStatusBar
      .systemStatusBar
      .statusItemWithLength(NSVariableStatusItemLength)
    @statusBar.highlightMode = true

    subscribeToConnectionEvents
  end

  def statusBarItemMenu=(menu)
    @statusBar.menu = menu
  end

  def connectionWillStartLoading(notification)
    startAnimation
  end

  def connectionDidFinishLoading(notification)
    stopAnimation
  end

  def showImage
    @statusBar.image = @menuIsActive ? alternateImage : image
  end

  # from timer
  def updateImage(timer)
    @statusBar.image = imageForFrame(@animationFrame)
    @animationFrame = (@animationFrame == 9) ? 1 : @animationFrame + 1
  end

  private

  def subscribeToConnectionEvents
    NSNotificationCenter.defaultCenter.addObserver self,
       selector:"connectionWillStartLoading:",
       name:"com.cimenu.CIMenu.semaphore.connectionWillStartLoading",
       object:nil

    NSNotificationCenter.defaultCenter.addObserver self,
       selector:"connectionDidFinishLoading:",
       name:"com.cimenu.CIMenu.semaphore.connectionDidFinishLoading",
       object:nil
  end

  def startAnimation
    unless @animationTimer
      @animationFrame = 0
      @animationTimer = NSTimer.scheduledTimerWithTimeInterval 0.02,
        target:self,
        selector:'updateImage:',
        userInfo:nil,
        repeats:true

      NSRunLoop.mainRunLoop.addTimer(@animationTimer, forMode:NSRunLoopCommonModes)
    end
  end

  def stopAnimation
    if @animationTimer
      @animationTimer.invalidate && @animationTimer = nil
    end
  end

  def imageForFrame(frame)
    NSImage.imageNamed "#{frame}.png"
  end
end

File: CIMenu/StatusBarItemController.swift
//
//  StatusBarItemController.swift
//  CIMenu
//
//  Created by Nickolay Abdrafikov on 09.06.14.
//  Copyright (c) 2014 Nickolay Abdrafikov. All rights reserved.
//

import Cocoa

class StatusBarItemController {
    let statusBar : NSStatusItem
    let image = NSImage(named:"gear_offline")
    let alternateImage = NSImage(named:"gear_clicked")

    let notificationCenter = NSNotificationCenter.defaultCenter()

    var menuIsActive = false
    var statusBarItemMenu : NSMenu {
        set(menu) {
            statusBar.menu = menu
        }
        get {
            return statusBar.menu
        }
    }
    
    init() {
        statusBar = NSStatusBar
            .systemStatusBar()
            .statusItemWithLength(-1) // linker error (bug) http://stackoverflow.com/a/24026327
//            .statusItemWithLength(NSVariableStatusItemLength)

        statusBar.highlightMode = true

        subscribeToConnectionEvents()
    }

    func showImage() {
        statusBar.image = menuIsActive ? alternateImage : image
    }

    @objc func startLoading() {
        println("startLoading")
    }

    @objc func stopLoading() {
        println("stopLoading")
    }

    private func subscribeToConnectionEvents() {
        notificationCenter.addObserver(self,
            selector: Selector("startLoading"),
            name: "org.cimenu.semaphore.connectionWillStartLoading",
            object: nil
        )

        notificationCenter.addObserver(self,
            selector: Selector("stopLoading"),
            name: "org.cimenu.semaphore.dataReceived",
            object: nil
        )

    }
}
